File: STRINGS/1802-number-of-students-unable-to-eat-lunch/number-of-students-unable-to-eat-lunch.java
class Solution {
    public int countStudents(int[] students, int[] sandwiches) {
        int student_0 = 0; // students preferring circular sandwiches
        int student_1 = 0; // students preferring square sandwiches

        // Count the number of students for each preference
        for (int student : students) {
            if (student == 1)
                student_1++;
            else
                student_0++;
        }

        // Iterate through sandwiches and check if any students are unable to eat
        for (int sandwich : sandwiches) {
            if ((sandwich == 0 && student_0 == 0) || (sandwich == 1 && student_1 == 0))
                break; // No more students can eat this type of sandwich
            else if (sandwich == 0)
                student_0--; // Reduce the count of available circular sandwiches
            else
                student_1--; // Reduce the count of available square sandwiches
        }

        // Sum up the remaining students who couldn't eat
        return student_0 + student_1;
    }
}

File: STRINGS/1802-number-of-students-unable-to-eat-lunch/number-of-students-unable-to-eat-lunch.cpp
class Solution {
public:
    int countStudents(vector<int>& students, vector<int>& sandwiches) {
        stack<int> sandwichStack;
        queue<int> studentQueue;

        for (int i = sandwiches.size() - 1; i >= 0; --i)
            sandwichStack.push(sandwiches[i]);
        for (int student : students)
            studentQueue.push(student);
        
        int unableToEat = 0;
        int studentsLeft = students.size();
        
        while (!studentQueue.empty()) {
            if (studentQueue.front() == sandwichStack.top()) {
                // If the student's preference matches the top sandwich, they can eat
                studentQueue.pop();
                sandwichStack.pop();
                studentsLeft--;
                unableToEat = 0; // Reset unableToEat counter
            } else {
                // If the student's preference doesn't match, move the student to the end of the queue
                int frontStudent = studentQueue.front();
                studentQueue.pop();
                studentQueue.push(frontStudent);
                unableToEat++;
            }
            
            // If all students have been through the queue without eating, break the loop
            if (unableToEat == studentsLeft)
                break;
        }
        
        return studentsLeft;
    }
};
